clients/admin-ui/src/features/datamap/SpatialDatamap.tsx

```diff
@@ -34,26 +37,34 @@ const useSpatialDatamap = (rows: Row<DatamapRow>[]) => {
         }, {} as Record<string, SystemNode>),
     [rows]
   );
-
   const data = useMemo(() => {
     let nodes: SystemNode[] = [];
-    let links: Link[] = [];
+    const links: Set<string> = new Set([]);
     if (datamapBySystem) {
       nodes = Object.values(datamapBySystem);
-      links = nodes.reduce(
-        (acc: Link[], system: SystemNode) => [
-          ...acc,
-          ...(system.dependencies
-            ?.filter((dependency) => datamapBySystem[dependency])
-            .map((dependency: string) => ({
+      nodes
+        .map<Link[]>((system) => [
+          ...system.ingress
+            .filter((ingress_system) => datamapBySystem[ingress_system])
+            .map((ingress_system) => ({
+              source: ingress_system,
+              target: system.id,
+            })),
+          ...system.egress
+            .filter((egress_system) => datamapBySystem[egress_system])
+            .map((egress_system) => ({
               source: system.id,
-              target: dependency,
-            })) ?? ([] as Link[])),
-        ],
-        []
-      );
+              target: egress_system,
+            })),
+        ])
+        .flatMap((link) => link)
+        .forEach((link) => links.add(JSON.stringify(link)));
```

Comment:
@allisonking The reason why you were seeing so many duplicate links is because the set was never working... 😅 It turns out that Sets in typescript don't work on objects (unlike Rust 🙄) so I need to convert each link into a string so it does that deduping then convert them back to an object.

Reply:
ah tricky!!! nice job catching that, the map is looking cleaner 👍